### ISSUE TITLE
Refactor GeoPandas as an extension array type in Pandas

### ISSUE THREAD

**Issue by mrocklin** [geopandas/geopandas#680]
Due to recent work by @TomAugspurger in https://github.com/pandas-dev/pandas/pull/19520 and elsewhere, it is now becoming easier to define extension array types and namespaces for Pandas.  This allows projects to add new column types (like geometries) and namespaces (like our geometry operations) to vanilla Pandas Series without subclassing DataFrame (which is difficult).

This poses a question for GeoPandas:

> Can we reduce this project to just a new array block type and some functions, dropping our subclasses?

I suspect that the answer is almost certainly "No, there is more within GeoPandas that would not easily fit within the new changes to Pandas", but I think that it might be an interesting exercise to list what the challenges here are, as they might inform future development within Pandas.  If something like this were to work then it might both reduce the maintenance burden on the GeoPandas devs, increase the exposure of the GeoPandas code due to a reduced switching cost, and further encourage extensions to play nicely with each other (like geo-cyber-pandas).

What are some operations within GeoPandas that would not work well if we were to drop the GeoDataFrame and GeoSeries classes?

**Comment by mrocklin** [comment]
@TomAugspurger do pandas.DataFrame objects collect namespaces from their blocks?  If I have a GeometryBlock or IPAddressBlock within a dataframe can that dataframe use the `.geo` or `.ip` namespaces?

**Comment by mrocklin** [comment]
@jorisvandenbossche brought up that we would have to track metadata and other attributes like spatial indices somewhere.  Previously these were on the GeoDataFrame or GeoSeries object, but now would have to be tracked on the block.

**Comment by mrocklin** [comment]
There are presumably various methods like `GeoDataFrame.plot` that would have to move within a namespace.

**Comment by TomAugspurger** [comment]
> do pandas.DataFrame objects collect namespaces from their blocks?

No, the accessor is registration is done at import-time of the library adding the accessor, and Series and DataFrame accessors are independent.

>  If I have a GeometryBlock or IPAddressBlock within a dataframe can that dataframe use the .geo or .ip namespaces?

If geopandas registered a `.geo` accessor, then all dataframes would have a `.geo` attribute. It's up to the accessor what to do when accessed. It could certainly return the geometry column so that this wouldn't be an API change.

>  If I have a GeometryBlock or IPAddressBlock

(just FYI, there's no `IPAddressBlock` and ideally `GeometryBlock` would go away. There will just be an `ExtensionBlock` with ipaddress data, or a `GeometryBlock` with (pointers to) geometry data.)

> There are presumably various methods like GeoDataFrame.plot that would have to move within a namespace.

Perhaps, though I have another refactor planned so that other systems can take over `.plot` if they implement the right interface.

**Comment by mrocklin** [comment]
Ah, so once I import geopandas *all* pandas.DataFrame objects will have a `.geo` accessor if I define one in the geopandas library.  Correct?

My guess is that the GeoPandas plot function might be sufficiently distinct that we would want to ensure that it is separate.  I don't know though.

**Comment by TomAugspurger** [comment]
>  Ah, so once I import geopandas all pandas.DataFrame objects will have a .geo accessor if I define one in the geopandas library.  Correct?

That's correct. Just like all series have a `.cat` attribute. We just raise an `AttributeError` if you access it on a non-categorical column.

In the same way, geopandas' dataframe accessor would raise an `AttributeError` if there wasn't actually a geometry column (if I understand the current functionality of `.geo`).

You're probably right about the plotting.

**Comment by martindurant** [comment]
An accessor can provide any number of features, right? So you can have `df.mycolumn.geo.plot(..)` with the current functionality. If you do `df.geo.whatever()`, does the method get the whole dataframe, and be able to see if there are one-or-more geo columns to work with?

**Comment by TomAugspurger** [comment]
> does the method get the whole dataframe

Yep, see http://pandas-docs.github.io/pandas-docs-travis/developer.html#registering-custom-accessors

The only restriction on the accessor is that its `__init__` looks like `def __init__(self, pandas_obj)`. From there, you're free to do whatever.

**Comment by martindurant** [comment]
Cool! Once your work is released, we should brainstorm data types that would be useful. I'll be fascinated to know what other vectorized functionality people come up with... but that discussion probably doens't belong here.

**Comment by jorisvandenbossche** [comment]
First, it's all about arrays, not blocks (the GeometryBlock will indeed be removed, once the extension array stuff is fully landed in pandas). 
And I have a branch that I keep in sync with Tom's work and changes our already existing GeometryArray into a proper extension array, which is working nicely.

Then, regardless of removing the subclasses or not, by using this extension mechanism, we will gain compatibility anyhow. Pandas will be able to better handle the geometry column, and it should not be converted to an object column anymore, even in normal DataFrames. (and eg, putting a IPAddress column in a GeoDataFrame will already be possible). 
So this is something we are going to do anyhow. The discussion is thus mainly: "can using accessors replace the GeoDataFrame/GeoSeries subclasses"

**Comment by jorisvandenbossche** [comment]
Then, about the actual discussion: I had already thought about it briefly, and had some reservations (both user API wise, as implementation wise). But it might be that I didn't think about it in enough detail. 
Eg I wasn't aware that there are also dataframe accessors (the pandas internal ones are all series/index accessors), that would indeed already solve some things. As we have not only special methods on the GeoSeries, but also on GeoDataFrame (`plot` as already mentioned, but also `to_file`, `to_crs`, `dissolve`, `explode`, ..).

Some random thoughts:

- We have currently the special case that a geometry column does not need to be called 'geometry', but `df.geometry` always returns the 'active' geometry column. When there is one geometry column, this could be inferred each time when a `.geo` accessor is called, but if you have multiple geometry columns, this will be a problem.
- Metadata: the `crs` attribute (apart from the geometry column name above). For subclasses pandas has some mechanisms to preserve this. So when dropping the subclasses, this would need to be moved to the GeometryArray level, but that's probably a good idea anyhow (even if not removing the subclasses). But that's something that needs to be explored with the ExtensionArray (currently we have no mechanism to preserve metadata on this level, although that's maybe not needed, and the `crs` could maybe also live on the dtype).
- Ideally it would be nice if `df.geometry` still returned the active geometry column, but we cannot use the dataframe-accessor for both returning the geometry column and for dataframe level methods (but, could use different names like `.geo` and `.geometry` ..). But even if `df.geometry` returned the geometry column, or just in the case that your geometry column is called 'geometry', you would need to do something like `df.geometry.geo.x` to access a specific geo attribute (since `df.geometry` will be a normal Series and no longer GeoSeries), which is not as nice as the current `df.geometry.x`. 
  Of course, we could also add all those attributes on the dataframe-accessor as well (and not only the series-accessor). But then we run into the multiple geometry columns problem.

Some user facing API changes would be:

- DataFrame methods like `plot`, `to_crs`, `to_file`, `dissolve`, .. will now be accessed through a `.geo` accessor
- When working with multiple geometry columns, the dataframe methods from above + the geopandas functions that accept a (Geo)DataFrame, will need a way to choose the correct geometry column.
- Currently also most geo methods/attributes are available on the GeoDataFrame, and not only on the GeoSeries, also for methods that only need the geometry column information (eg `df.intersects(p)` is the same as `df.geometry.intersects(p)`. I personally don't really like this fact, but in practice many people will use it, and so this will also be a change.

**Comment by mrocklin** [comment]
> When working with multiple geometry columns, the dataframe methods from above + the geopandas functions that accept a (Geo)DataFrame, will need a way to choose the correct geometry column.

Hrm, that's an interesting challenge.  

> Currently also most geo methods/attributes are available on the GeoDataFrame, and not only on the GeoSeries, also for methods that only need the geometry column information (eg df.intersects(p) is the same as df.geometry.intersects(p). I personally don't really like this fact, but in practice many people will use it, and so this will also be a change.

I think we can keep this if we want to?  From what I understand we can register an accessor namespace on both Series and DataFrame.  

Alternatively, if we're going to be making a lot of changes and breaking user code then we might as well break some more.  My hope would be that the performance increase is enough to motivate users to change their code.  This is a gamble though.

**Comment by jorisvandenbossche** [comment]
> I think we can keep this if we want to? From what I understand we can register an accessor namespace on both Series and DataFrame.

Using accessors, we cannot keep the fact that now `df.intersects(p)` works, which will need to be changed to eg `df.geo.intersects(p)` (indeed, using the dataframe accessor)

**Comment by mrocklin** [comment]
Right, agreed.  Sorry I thought you were referring to the need to do `df.my_geo_column.geo.intersects(p)` rather than `df.geo.intersects(p)`

**Comment by mangecoeur** [comment]
Just wondering if this had been implemented? I was wondering if it would make it possible to use GeoSeries without having to convert uses of pandas DataFrame to GeoDataFrame? This would make it more similar to Postgis, where geo data types are just another column type (and make it easier to mix geo and non-geo data moving back and forth between python and postgis)

**Comment by jorisvandenbossche** [comment]
@mangecoeur For the ExtensionArray part, there are open PRs. See https://github.com/geopandas/geopandas/pull/835 (for master) and https://github.com/geopandas/geopandas/pull/701 (for cython branch). Those PRs are a bit stale, but I am going to get back to those somewhere in April (unless sombody already wants to take them over now).

But, that is only the aspect of using an ExtensionArray to store the geometries instead of a numpy object dtype array. That is something that will happen for sure.

Having such geometry ExtensionArray will make it easier to store the geometry data just in a normal pandas DataFrame/Series. But the main discussion point in this issue, is how to expose geospatial functionality. Because typically, you don't want to only store the geometry data in a dataframe, you also want to make use of the geospatial functionality (like all the `ST_` functions in postgis). 
That is what the current GeoDataFrame/GeoSeries provide (giving access to geospatial functionality). This could be (partly) replaced by using a `geo` accessor on a plain DataFrame/Series. But, I am personally not yet fully convinced that it will give an as nice interface for users as the current subclasses (see discussion above). So I would be very interested to get feedback on this aspect.

> and make it easier to mix geo and non-geo data moving back and forth between python and postgis

In what way would it be easier?

**Comment by mangecoeur** [comment]
> In what way would it be easier?

When doing data analytics with Postgres/PostGIS, often you have a mix of numerical tables and GIS tables which could have several geometry type columns. This means that if you are pulling/pushing data from Postgres, you have to decide whether to pull it as a pandas dataframe or a geopandas dataframe. Maybe you start working with purely numerical data then you want to add a GIS aspect -  you have to convert part or all of your analysis to use geopandas frames.  Now you have to keep track of what tables are GIS-enabled and which aren't, and also which operations are going to produce GeoDataFrames vs normal ones. 

If instead the GIS data columns were simply an extension type to the regular column, you can just always use pandas DFs with as many geo columns as you like (and any other extension columns).

With regards to the API, the `geo` accessor is a nice and consistent pattern which doesn't assume a single geometry per table (this might be a common convention but from a pure data processing perspective there is no reason why a table should only have one geometry). I would put 'special' functions that need to operate on a dataframe rather than a single geometry column as top-level functions (`gpd.whatever()` ).

This also solves the 'round trip' problem, whereby taking a `GeoDataFrame` apart and putting it  back together results in a different dataframe type. See example using Natural Earth Admin0

```python
type(COUNTRIES)
>>> geopandas.geodataframe.GeoDataFrame
geom = COUNTRIES['geometry']
data = COUNTRIES[['sovereignt','sov_a3']].copy()

type(geom)
>>> geopandas.geoseries.GeoSeries

type(data)
>>> pandas.core.frame.DataFrame

data['geometry'] = geom

type(data)
>>> pandas.core.frame.DataFrame

```

Interestingly you get a different result if you `drop()` the geometry column instead of selecting column subsets - I imagine there are a lot of cases

**Comment by caspervdw** [comment]
The performance boost described in your blogs @jorisvandenbossche  would for me absolutely outweigh the API changes. Most API changes are actually rather small (adding .geo. at many places) so migrating to the new API would not take too much effort for most users.

The API is of course a matter of application. As a regular pandas user I was a bit surprised that I could operate on the GeoSeries by calling methods on the GeoDataframe. This confused me especially when dealing with multiple geometry columns. Removing this functionality, forcing users to access the GeoSeries and then call the ST_ operation, would be less confusing to the pandas user.

But again, it is just how you use the geodataframe. If you always have precisely one geometry column the current API makes perfect sense. In absence of the knowledge of how people use geopandas I would prefer the “explicit is better that implicit” approach, which is forcing users to specify the column(s) they want to operate on.

One other idea could be putting most functionality in the ExtensionArray, but keeping the GeoDataFrame as a thin wrapper to (sort of) maintain the current API?

**Comment by maxhully** [comment]
What about releasing the GeometryArray as a separate "core" package that GeoPandas depends on/enriches with additional behaviors? I'm imagining that the higher-level functionality like plotting, spatial joins, and overlay/dissolve, for example, would all belong in the full GeoPandas package, while the core package would focus solely on nailing the lower-level GeometryArray implementation. (If it seems useful, you could include basic shapely operations through an optional `.geo` accessor in the core package and then offer the current API through the flagship GeoPandas package, similar to caspervdw's suggestion in the previous comment.)

That could be useful for other libraries that want to be able to work with geospatial data using pandas, but don't want to depend on all of geopandas' dependencies---since the conda-forge distribution has to include all of the optional dependencies. Conceivably this core package could depend solely on shapely and pandas, which could alleviate/offer a workaround for anyone dealing with conflicting versions lower down in the stack.

**Comment by martinfleis** [comment]
Implemented in 0.6. Thank you all!

**Comment by jorisvandenbossche** [comment]
So in 0.6 the ExtensionArray was indeed implemented, but in this issue we also have been discussing using this as as in pandas DataFrames, without the subclasses that GeoPandas provides right now (using the accessors). So let's keep this open for now.

However, now that the extension array is released, one can experiment with the accessor idea.

**Comment by kukushking** [comment]
Hi folks, looking at this refactoring it seems you're half way done. I would like to know if it is already possible to use geometric data types as an extension of regular pandas data frames types e.g. I would like to:
```
import pandas as pd
from geopandas.types import Polygon

df = pd.DataFrame([...], dtypes={"geo": Polygon})
```

**Comment by m-richards** [comment]
Hi @kukushking, this is quite an old issue and as mentioned in the thread, ExtensionArrays have been implemented for a while now (perhaps the title of this issue no longer reflects the discussion within it).

 It's already possible to use geometry data in ordinary dataframes (although this is not exactly the main supported use case):

```python
import pandas as pd
import geopandas as gpd
s = gpd.GeoSeries.from_xy([1,2], [3,4])

df =pd.DataFrame({"a":[1,2], "b":s})
print(df.dtypes)
a       int64
b    geometry
dtype: object
```

This however isn't imposing a specific Polygon / Point column dtype. If you'd like to elaborate on your use case feel free to open a fresh issue to discuss further.

**Comment by kukushking** [comment]
@m-richards Awesome! If you have any documentation on using these types with Pandas I would really appreciate it.

The use case I'm looking into is being able to query geospatial data using AWS Athena [#1561](https://github.com/aws/aws-sdk-pandas/issues/1561) - in which case I would need to convert Athena geospatial types (point, line, polygon, multiline, multipolygon) to equivalent Geopandas types and return the data in a Pandas data frame.

**Comment by m-richards** [comment]
> @m-richards Awesome! If you have any documentation on using these types with Pandas I would really appreciate it.
> 
> The use case I'm looking into is being able to query geospatial data using AWS Athena [#1561](https://github.com/aws/aws-sdk-pandas/issues/1561) - in which case I would need to convert Athena geospatial types (point, line, polygon, multiline, multipolygon) to equivalent Geopandas types and return the data in a Pandas data frame.

I guess the closest analogue in terms of functionality already in geopandas is probably the postgres IO (saying this without knowing much about Athena): https://github.com/geopandas/geopandas/blob/4b72fb417a4e31ad4ffde7df01927faf54ac362b/geopandas/io/sql.py#L100
